src/main/ngapp/accio/src/app/editor/magic-wand.service.ts

```diff
@@ -0,0 +1,159 @@
+      let x: number = coord[0];
+      let y: number = coord[1];
+
+      // Operational part of while-loop
+      mask.add(this.coordToDataArrayIndicies(x, y, imgData.width)[0]);
+
+      // Loop part of while-loop
+
+      // Get coords of adjacent pixels
+      let leftPixelCoord: Array<number> = [x - 1, y];
+      let rightPixelCoord: Array<number> = [x + 1, y];
+      let upPixelCoord: Array<number> = [x, y - 1];
+      let downPixelCoord: Array<number> = [x, y + 1];
+      // Add coords of adjacent pixels to the heap
+      let isStillMask: boolean = this.getIsStillMask(originalPixel, imgData, leftPixelCoord, tolerance, visited);
+      // Check left pixel
+      if (isStillMask) {
+        visit.push(leftPixelCoord);
+        visitSize += 1;
+      }
+      isStillMask = this.getIsStillMask(originalPixel, imgData, rightPixelCoord, tolerance, visited);
+      // Check right pixel
+      if (isStillMask) {
+        visit.push(rightPixelCoord);
+        visitSize += 1;
+      }
+      isStillMask = this.getIsStillMask(originalPixel, imgData, upPixelCoord, tolerance, visited);
+      // Check up pixel
+      if (isStillMask) {
+        visit.push(upPixelCoord);
+        visitSize += 1;
+      }
+      isStillMask = this.getIsStillMask(originalPixel, imgData, downPixelCoord, tolerance, visited);
+      // Check down pixel
+      if (isStillMask) {
+        visit.push(downPixelCoord);
+        visitSize += 1;
+      }
+    }  // end of while loop
+
+    return mask;
+  }
+
+  /** Judge current pixel's RGBA against original pixel's RGBA to see if it
+   *  can still be part of the mask (using tolerance criteria)
+   */
+  getIsStillMask(originalPixel: Array<number>, imgData: ImageData,
+                         pixelCoord: Array<number>, tolerance: number,
+                         visited: Set<number>): boolean {
+    let isStillMask: boolean = true;
+
+    // Preface; check if pixel is valid (indexing errs and repeat values)
+    let isValid: boolean = this.getIsValid(imgData, pixelCoord, visited);
+    if (!isValid) {
+      // Automatically not in mask b/c failed vailidity test
+      return !isStillMask;
+    }
+
+    // Get array of attributes of current pixel
+    let curX: number = pixelCoord[0];
+    let curY: number = pixelCoord[1];
+    let curPixel: Array<number> = this.dataArrayToRGBA(imgData, curX, curY);
+
+    // All attributes of the pixel (R,G,B, and A) must be within tolerance level
+    for (let i = 0; i < 4; i++) {
+      let upperTolerance: boolean = curPixel[i] > originalPixel[i] + tolerance;
+      let lowerTolerance: boolean = curPixel[i] < originalPixel[i] - tolerance;
+      let failsTolerance: boolean = upperTolerance || lowerTolerance;
+      if (failsTolerance) {
+        return !isStillMask;
+      }
+    }
+
+    return isStillMask;
+  }
+
+  // Check if @pixelCoord is in bounds and makes sure it's not a repeat coord
+  getIsValid(imgData: ImageData, pixelCoord: Array<number>, visited: Set<number>): boolean {
```

Comment:
Seems expensive to pass the entire ImageData object for every call of `getIsValid`. Looks like only the dimensions are needed for this function. Similarly for other functions, try to minimize the amount of information passed around.

Reply:
Change request addressed in the commits below.

On a side note: can we also discuss the cost of of switching from imgData >> imgData.width, imgData.height as args for the method for the purposes of future code design? 
First, my assumption was that having less arguments as parameters for methods would maintain cleaner code as far as readability, as well as inputting broader data types as arguments would keep the code more flexible if it needs to be altered down the line.
Second, I was wondering if the difference between passing in ImageData as a param as opposed to passing in imaData.width and imgData.height as params are actually negligible in terms of operational expense. 
For example, by passing in imgData: ImageData as a parameter, I thought we are just passing in a single address in memory that holds the value of an address as a variable reference to the actual ImageData object, and thus is approx. the same as passing around imgData.width and imgData.height, which are two separate addresses in memory that hold similar things respectively. Based on this assumption, how I understand it is that the difference is equivalent to accessing 1 extra memory address in the code per use of getIsValid(), which leads to O(n + 1) is O(n). 
I'm also mainly asking because this might be a good opportunity for me to learn from you, because my impression is that you might be more familiar with C than me, which gives me the assumption that you might also understand the little nuances of costs of operations more than me.

Reply:
I was unclear about when JS passes by reference vs value. LGTM

Reply:
In JS objects and arrays are passed by reference and otherwise value.  By being implicit it leads to this uncertainty.